snapcraft/extensions/_extension.py

```diff
@@ -0,0 +1,118 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2018-2019 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import abc
+import os
+from typing import Any, Dict, Optional, Tuple, final
+
+from craft_cli import emit
+
+from snapcraft import errors
+
+
+class Extension(metaclass=abc.ABCMeta):
+    """Extension is the class from which all extensions inherit.
+
+    Extensions have the ability to add snippets to apps, parts, and indeed add new parts
+    to a given snapcraft.yaml.
+
+    :param extension_name: The name of the extension.
```

Comment:
`extension_name` is not declared in `__init__()` below. (This is easier to spot when `__init__()` is the first method declared in the class)

Reply:
fixed